# Tasks

This notebook will hold my solutions to the Tasks Assessment for Machine Learning and Statistics 2020.
(Task 1 is the only task currently available.)

The author is Angela Carpenter

## Task 1:
**Write a Python function called sqrt2 that calculates and prints to the screen the square root of 2 to 100 decimal places. Your code should not depend on any module from the standard library or otherwise. You should research the task first and include references and a description of your algorithm.**
***

The square root of two is the number that when multiplied by itself gives the number two.
It is a geometric number as it refers to the perfect square where each side of the square is of length 1. The length of the diagonal of a perfect square is the square root of the number 2 (Pythagoras theorem).

I do recall looking up square roots for problem 7 of the Problem set of Programming and Scripting module. This problem involved writing a program to take a positive floating point number as input and to output an approximation of it's square root.

First to Wikipedia for any information on why this particular number is noteworthy.
According to [1][Wikipedia](https://en.wikipedia.org/wiki/Square_root_of_2)., the square root of two was probably the first number known to be irrational.
>*The square root of 2, or the one-half power of 2, written in mathematics as 
$2{\sqrt {2}}$ or $2^{{1/2}}$ is the positive algebraic number that, when multiplied by itself, equals the number 2.*

>Technically, it must be called the principal square root of 2, to distinguish it from the negative number with the same property.
[1][Wikipedia](https://en.wikipedia.org/wiki/Square_root_of_2).

There is a lot on the Wikipedia page so I will come back to it.


[2][mathsisfun](https://www.mathsisfun.com/numbers/square-root-2-irrational.html) has a nice easy explanation of why the the square root of two is an irrational number and can never be written as a fraction.

### The square root of two.
Here I look at what the square root of two should look like.
There is a Python function for calculating square roots in the maths module which is part of the Python Standard Library. See [math module](https://docs.python.org/3/library/math.html?highlight=math#module-math).
This task should not use any libraries but the [math.sqrt](https://docs.python.org/3/library/math.html?highlight=math#math.sqrt) function could be used to check the results of my code.
`math.sqrt(x)` calculates the square root of whatever is passed in as `x`.
This task however is to calculate the square root of the number 2 to 100 decimal places. 

$$\sqrt{2}$$

The square root of a number when multiplied by itself should give the number so here I check that by multiplying the result of `math.sqrt(2)` by `math.sqrt(2)`.

In [11]:
# import math module
import math
# using the sqrt function
print(math.sqrt(2))
print(math.sqrt(2)*math.sqrt(2))

1.4142135623730951
2.0000000000000004


In [10]:
import numpy as np
print(np.sqrt(2))
print(np.sqrt(2)*np.sqrt(2))
np.sqrt(2)*np.sqrt(2)==2

1.4142135623730951
2.0000000000000004


False

In [9]:
# just seeing if the numpy and math square roots are the exact same
math.sqrt(2) == np.sqrt(2)
math.sqrt(2)*math.sqrt(2) == np.sqrt(2)*np.sqrt(2)


True

The results above f multiplying the square root of two returned from both the `math.sqrt` function and the `np.sqrt` funciton is 2.0000000000000004 and I guess this demonstrates the problem with performing calculations on the irrational number square root of two!

## Binary representation of integers and floats.

#### References used here:
- [1][Floating Point Arithmetic: Issues and Limitations](https://docs.python.org/3/tutorial/floatingpoint.html) 
 
- [2] Lecture Notes from Machine Learning and Statistics, Lecturer Dr. Ian McLoughlin
The lecture on data types of this module looked at the binary representation of integers and floats.





Floating point numbers are stored with a finite number of digits so there is a finite limit on how many decimal places they can actually store. 
- Real numbers such as $\sqrt{2}$ and $\pi$ have an infinite number of places after the decimal point. Computers cannot really store such real numbers as there is only a finite number of bits available to store the number.
- Floating point numbers are actually stored using integers in two parts, for the whole number part before the decimal point and the part of the whole after the decimal point. (Think it might be called the characteristics for the whole number and the mantissa for the part after the decimal point).

- The set of $\mathbb{N}$ Natural numbers are an infinite set but it is the smallest infinite set there is and is known as countable infinite. The set of measuring numbers are all the numbers on the number line that are used for quantifying and measuring things and there is an infinite amount of such numbers between any two discrete points on the number line.
 
- Computer programs are made up of binary strings of 1's and 0's. Binary strings can be converted into integers. Every computer program can be converted into one of the counting numbers from the set of $\mathbb{N}$ . According to Turing, all computer programs are countable which means they can be put into one-to-one correspondance with the natural numbers but Turing also realised that not all numbers can be calculated or computed.

-  A rational number can be expressed as the ratio of 2 whole numbers. It can be stored as an array of two numbers. A number is rational only if it's decimal expansion becomes periodic.

- The square root of two is an irrational number and therefore cannot be expressed in the ratio of two numbers in their most simplified form.

- The implications of this for computing is that the square root of two can never be stored. An approximation of its value can be stored though. This is the same for the irrational number $pi$. Irrational numbers are often stored as procedures which are used to calculate but not store the number.
- Caution is always required when calculation statistics or performing analytics on irrational numbers. It is possible to work theoretically or to use approximations but rounding errors could interfere will the calculations.

### Floating Point Arithmetic: Issues and Limitations
This section here is based on the tutorial in the python docs at chapter 15.
[Chapter 5 of the Python Tutorial](https://docs.python.org/3/tutorial/floatingpoint.html) looks at the issues and limitations associated with floating point arithmetic. It illustrates how the decimal fraction 0.125 and the binary fraction 0.001 have identical values:

- Floating-point numbers are represented in computer hardware as base 2 (binary) fractions
- The decimal fraction 0.125 has value $\frac{1}{10}$ + $\frac{2}{100}$ + $\frac{5}{1000}$
- The binary fraction 0.001 has value $\frac{0}{2}$ + $\frac{0}{4}$ + $\frac{1}{8}$
The only real difference here is that the decimal fraction 0.125 is written in base 10 fractional notation while the binary fraction 0.001 above is written in base 2 notation.

>Unfortunately, most decimal fractions cannot be represented exactly as binary fractions. A consequence is that, in general, the decimal floating-point numbers you enter are only approximated by the binary floating-point numbers actually stored in the machine.

(It demonstrates how the fraction 1/3 can be approximated as a base 10 fraction as  0.3 or 0.33 or 0.333. No matter how many 3's you use the result will never actually be $\frac[1}{3}$ but it will be increasingly a better approximation of $\frac[1}{3}$)

The decimal value 0.1 cannot be represented exactly as a base 2 fraction because in base 2, $\frac{1}{10}$ is the infinitely repeating fraction
0.0001100110011001100110011001100110011001100110011...
If you stop at any  finite number of bits you get an approximation. 
On most machines today floats are approximated using a binary fraction. 
- The numerator uses the first 53 bits starting with the most significant bit
- The denominator as a power of 2
The binary fraction for the decimal 0.1 is `3602879701896397 / 2 ** 55` which is close to the true value of 1/10 but is not exactly equal to the true value of 0.1.
We don't notice that this is not the true value and only an approximation because of the way values are displayed. Python (and other languages) only prints *a decimal approximation to the to the true decimal value of the binary approximation stored by the machine* 

- The true decimal value of the binary approximation stored for 0.1 is actually `0.1000000000000000055511151231257827021181583404541015625`. It is printed as 0.1 though.
- Python displays a rounded value instead of all the digits as in most situations this is enough.
- The important point is that even though what we see printed (0.01) looks like the exact value of $\frac{1}{10}$, the actual value that is stored is the nearest representable binary fraction.

- The tutorial here also illustrates how there are many different decimal numbers which actually share the same nearest approximate binary fraction. The numbers `0.1`  and `0.10000000000000001` and `0.1000000000000000055511151231257827021181583404541015625` are all approximated by `3602879701896397 / 2 ** 55` . These decimal values all have the same approximation.

- String formatting can be used to produce a limited number of significant digits.
For example `format(math.pi, '.12g')` gives 12 significant digits while `format(math.pi, '.2f')` gives 2  digits after the point.
However this is simply rounding the **display** of the true machine value.

See [The Perils of Floating Point](http://www.lahey.com/float.htm)
See [str.format](https://docs.python.org/3/library/stdtypes.html#str.format)

The errors in Python float operations come from the floating-point hardware.
>The errors in Python float operations are inherited from the floating-point hardware, and on most machines are on the order of no more than 1 part in `2**53` per operation. That’s more than adequate for most tasks, but you do need to keep in mind that it’s not decimal arithmetic and that every float operation can suffer a new rounding error.

- The [decimal module](https://docs.python.org/3/library/decimal.html#module-decimal) implements decimal arithmetic suitable for accounting applications and high-precision applications (but not for this task!).
- The [fractions module](https://docs.python.org/3/library/fractions.html#module-fractions)  implements arithmetic based on rational numbers (so the numbers like 1/3 can be represented exactly).
- NumPy and [scipy](https://scipy.org) packages are recommended for mathematical and statistical operations.
- Python also has a [float.as_integer_ratio()](https://docs.python.org/3/library/stdtypes.html#float.as_integer_ratio) method which expresses the value of a float as a fraction. This can help when you want to know the exact value as a float. As the ratio is exact, it can be used to losslessly recreate the original value.


- **Representation error** refers to the fact that most decimal fractions cannot be represented exactly as binary (base 2) fractions in Python and other languages and therefore won’t display the exact decimal number you expect.
- Most machines use IEEE-754 floating point arithmetic.
- Almost all platforms map Python floats to IEEE-754 “double precision”.
- 754 doubles contain 53 bits of precision. 
The computer strives to convert 0.1 to the closest fraction that it can of the form $J/2**N$ where J is an integer containing exactly 53 bits.


- See [IEEE 754](https://en.wikipedia.org/wiki/IEEE_754) and [IEEE Standard 754 Floating Point Numbers](https://www.geeksforgeeks.org/ieee-standard-754-floating-point-numbers/) article on <geeksforgeeks.com>.
- IEEE 754 has 3 basic components:
    * The sign of the mantissa where `0` represents a positive number while `1` represents a negative number.
    * The biased exponent
    * The Normalised Mantissa
    
### Come back to this!

#### Using `float.as_integer_ratio()`.

See [float.as_integer_ratio()](https://docs.python.org/3/library/stdtypes.html#float.as_integer_ratio) method which expresses the value of a float as a fraction. This can help when you want to know the exact value as a float. As the ratio is exact, it can be used to losslessly recreate the original value.

This function is a built-in Python function. 

In [61]:
x = np.sqrt(2)
x.as_integer_ratio()

(6369051672525773, 4503599627370496)

#### Format Strings

In [58]:
print(format(math.pi, '.12g'))
print(format(math.pi, '.2f'))
print(repr(math.pi))

3.14159265359
3.14
3.141592653589793


In [43]:
x = 1/10

In [53]:
format(x,'.19g')

'0.1000000000000000056'

- The built-in [repr](https://docs.python.org/3/library/functions.html#repr) function returns a string containing a printable representation of the object.
- See also the [eval](https://docs.python.org/3/library/functions.html#eval) built-in function.

In [39]:
print(repr(0.1))
print(repr(0.10000000000000001))
print(repr(0.1000000000000000055511151231257827021181583404541015625))

0.1
0.1
0.1


In [2]:
import math
print(repr(math.sqrt(2)))

1.4142135623730951


In [4]:
x = 0.1000000000000000055511151231257827021181583404541015625
print(eval(repr(x)) == x)
# 2 to the power of 55
print(2**55)
# True
print(3602879701896397/ (2**55))
print((3602879701896397/ (2**55)) == 0.1)

True
36028797018963968
0.1
True


#### Binary representation
The `bin` function can be applied to any integer to get the binary representation of an integer.
The `struct` module can be used to get the binary representation of a float.

In [25]:
import struct
f = 1/3
bin(struct.unpack('!i',struct.pack('!f', f))[0])

'0b111110101010101010101010101011'

In [29]:
f = 0.1
bin(struct.unpack('!i',struct.pack('!f', f))[0])


'0b111101110011001100110011001101'

---
## Calculating the Square root of two (from scratch!)


For the programming and scripting problem set I used Newton's method to get the estimate of the square root of the number, following the tutorial provide in the Tour of Go tutorial. So I'll look there again. 
Here are some of the resources I used then which may be useful here.
- [wikipedia](https://en.wikipedia.org/wiki/Newton%27s_method)
- [math.stackexchange](https://math.stackexchange.com/questions/350740/why-does-newtons-method-work)
- [mathinsight](https://mathinsight.org/newtons_method_refresher)

Newtons Method can be used to get an estimate of the square root of a number. 
Using Newtons Method, regardless of the starting estimate, once the square of the estimate gets close enough to the actual number itself, then that is a good approximation for the square root. An initial estimate is used for the square root.



In [9]:
from decimal import *
getcontext().prec = 28
Decimal(math.sqrt(2)) 

Decimal('1.4142135623730951454746218587388284504413604736328125')

In [10]:
#1.4142135623730951454746218587388284504413604736328125
num = 1.4142135623730951454746218587388284504413604736328125
squared =num*num
math.sqrt(squared)==num

True

## Methods for Calculating square roots
Finding the square root of a number is the inverse of finding the square of a number. The square of a number is the number times itself. The perfect squares are the squares of the whole numbers. The square root of a number is the number that when multiplied by itself gives the number.

>*The square roots of the perfect squares (e.g., 0, 1, 4, 9, 16) are integers. In all other cases, the square roots of positive integers are irrational numbers, and hence have non-repeating decimals in their decimal representations.* [2]  <https://en.wikipedia.org/wiki/Square_root>


Getting the square root of numbers that are not perfect squares is not as straightforward.
You have to look at the square roots of perfect squares that are close to your number in question. So if you are looking at the square root of 5 you would consider the square root of 4 (which is 2) and 9 (which is 3). Therefore the square root of 5 must lie between 2 and 3. Similarly for the square root of 10, look at the square root of the perfect square 9 (which is 3) and 16 (which is 4). So the square root of 10 must lie between 3 and 4.
Looking at the square roots of the nearest perfect squares on either side of the number in question will be the starting point for finding the square root of a non-perfect square. It involves some guesswork.

As mentioned earlier, the square root of two is considered a geometric number as it refers to the perfect square where each side of the square is of length 1. The length of the diagonal of a perfect square is the square root of the number 2 (Pythagoras theorem). But the number two itself is not a perfect square.



In [13]:
# The square root of 5 must be between the two perfects squares 4 and 9
print("Looking for square root of 5:")
print(np.sqrt(4))
print(np.sqrt(5))
print(np.sqrt(9))
# The square root of 10 must be between the two perfect squares 9 and 16
print("\nLooking for square root of 10:")
print(np.sqrt(9))
print(np.sqrt(10))
print(np.sqrt(16))

Looking for square root of 5:
2.0
2.23606797749979
3.0

Looking for square root of 10:
3.0
3.1622776601683795
4.0


Here is the code using Newton's Method adapted from the tour of go that I used for the square root in the programming and scripting problem set. (problem 7)
It was adapted from the example in the Tour of Go tutorial.

In [14]:
# The code here using Newtons Method is adapted from  https://tour.golang.org/flowcomtrol/8
x= 2
guess = 1
# while the absolute difference between the guess squared and the number in question is greater than 0.01
while abs((guess * guess) - x )> 0.000000001:
        # update the guess using the formula 
    guess -=((guess * guess)-x)/(2 * guess) 
    print(guess)
    # the approximation for the square root will be the final value of estimate from the  while loop
 # assign the final estimate to sqroot   
sqroot = guess  
sqroot

1.5
1.4166666666666667
1.4142156862745099
1.4142135623746899


1.4142135623746899

### How to calculate the square root of number without a computer!
There are several websites which show how to get the square roots of numbers in the way I would have learnt back in school. Finding the square root of a number that is not a whole number begins with some guesswork. You begin by looking at the nearest perfect squares on both sides of the number. I will work through an example here to ilustrate.

- https://en.wikipedia.org/wiki/Square_root

- https://www.homeschoolmath.net/teaching/square-root-algorithm.php

- http://www.math.com/school/subject1/lessons/S1U1L9DP.html


1. Find two perfect square roots that the number in question lies between. 
    * For example if you want to find the square root of 10, you are looking at the square root of 9 which is 3 and square root of 16 which is 4.
2. Take one of those square roots as an estimate and divide the number in question by it.
    * The square root of 10 must lie between 3 and 4
    * pick either 3 or 4 as the estimate
    * Divide number in question by the starting estimate (3 or 4)
        * 10 / 3 = 3.3333
        * 10 / 4 = 2.5
 
3. Take the average of the result of step 2 above and the current guess

    *  (3 + 3.3333)/2 = 3.16666667
    
4. Use result from previous step as the new estimate and repeat these two steps until you get your answer.
    * new guess = 3.166667
    * 10 / 3.166667 = 3.1578944
    * (3.166667 + 3.1578944)/2 = 3.1622807
    
    * next guess. = 3.1622807
    * 10 / 3.162280 = 3.16227532033849
    * (3.162280 + 3.16227532033849)/2
    * 3.162277660169245
    

- `np.sqrt(10)` will give 3.1622776601683795

Make a guess of what the root of the number in question could be based on the square root of the nearest perfect squares on either side.  Add the result of this to the number and question and divide by two to get the average as the new starting guess. Keep repeating

- The square root of 10 truncuated to 50 decimal places is 3.16227766016837933199889354443271853371955513932521 

- The square root of 5 truncuated to 50 decimal placecs is 2.23606797749978969640917366873127623544061835961152
- The square root of 2 truncuated to 50 decimal places is 1.41421356237309504880168872420969807856967187537694
See (https://en.wikipedia.org/wiki/Square_root).


### Applying the above to get the square root of 2.

$1^2=1$ and $2^2=4$, therefore the square root of 2 must fall between 1 and 2.
Use 1 as the starting guess.  (The square root of two is not two.) 

1. Find two perfect square roots that the number in question lies between. 
       * Take either `1` or `2`
2. Take one of those square roots as an estimate and divide the number in question by it.
    * The square root of `2` must lie between `1` and `2`
    * pick either `1` or `2` as the estimate
    * Divide number in question by the chosen estimate (`1` or `2`)
        * 2 divided by 1 = 2
        * (or 2 divided by 2 = 1)
 
3. Take the average of the result of step 2 above and the current guess

    *  (2 + 1)/2 = 1.5
    
4. Use result from previous step as the new estimate and repeat these two steps until you get your answer.
    - new guess = 1.5
        * 2 / 1.5 = 1.33333333
        * (1.5 +1.33333333 )/2 = 1.4166666650000002
    
    - next guess = 1.4166666650000002
        * 2 / 1.4166666650000002 = 1.4117647075432525
        * (1.4166666650000002 + 1.4117647075432525)/2 = 1.4142156862716262
        
    - next guess = 1.4142156862716262
         * 2/ 1.4142156862716262 = 1.4142114384777535
         * (1.4142156862716262 + 1.4142114384777535)/2 = 1.4142135623746899
         
    - next guess = 1.4142135623746899
    
    and so on.......
    
    

Putting the above algorithm into code...

In [73]:
# x is the number 2  that we want the square root of
#x = 2**10
x = 2
# guess is based on the square root of the nearest perfect square
#guess = 1**10
guess =1

# while the difference between 2 and the calculated square root is greater than a specified amount
while abs(x -(guess * guess)) > 0.00000000000001:
    
    y = x / guess
    ave = (guess + y)/2
    guess = ave
print(guess)

result = ('%.60f' %guess)
print(result)
print(type(result))
print('%.60g' %guess)

1.414213562373095
1.414213562373094923430016933707520365715026855468750000000000
<class 'str'>
1.41421356237309492343001693370752036571502685546875


In [74]:
# x is the number 2  that we want the square root of
#x = 2**10
x = 2 *(10**100)
# guess is based on the square root of the nearest perfect square
#guess = 1**10
guess =1

# while the difference between 2 and the calculated square root is greater than a specified amount
while abs(x -(guess * guess)) > 0.00000000000001:
    
    y = x / guess
    ave = (guess + y)/2
    guess = ave
print(guess)

result = ('%.60f' %guess)
print(result)
print(type(result))
print('%.60g' %guess)

1.414213562373095e+50
141421356237309502300937332480540322869392772694016.000000000000000000000000000000000000000000000000000000000000
<class 'str'>
141421356237309502300937332480540322869392772694016


In [75]:
format_guess = format(guess,'.100g')
print(format_guess)

141421356237309502300937332480540322869392772694016


In [2]:
guess **2

1.9999999999999996

In [27]:
print(2**100)
print(10**100)
(10**100)//(2**100)

1267650600228229401496703205376
10000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000


7888609052210118054117285652827862296732064351090230047702789306640625

### Putting the above into a function.
The initial guess can be set to either the number in question or 1. 

In [72]:
# define the function
def my_sqrt(x):
    """
    takes  one input which is the number to calculate the square root of
    
    """
    guess = x 
    while abs(x -(guess * guess)) > 0.00000000000001:
        y = x / guess
        ave = (guess + y)/2
        guess = ave
    return(guess)
        
    

In [ ]:
# try the function on the number 2
my_sqrt(200)

### Rounding / Accuracy etc

At some stage the results gets rounded which is the main problem here! From Wikipedia the square root of 2 to 50 decimal places is `1.41421356237309504880168872420969807856967187537694`
The square root of 2 using `np.sqrt` to 50 decimal places is 1.41421356237309514547462185873882845044136047363281.
The power of two is the opposite of getting the square root.
The `**` operator can be used to get powers in Python.


In [13]:
1.41421356237309514547462185873882845044136047363281**2

2.0000000000000004

In [19]:
print(2-(1.414213562373095 * 1.414213562373095))
print(2-(1.41421356237309514547462185873882845044136047363281 * 1.41421356237309514547462185873882845044136047363281))

4.440892098500626e-16
-4.440892098500626e-16


In [21]:
import math
root2 = math.sqrt(2)
print(" The square root of two is {:.50f}".format(root2))

root2 = math.sqrt(200)
print(" The square root of two is {:.50f}".format(root2))

 The square root of two is 1.41421356237309514547462185873882845044136047363281
 The square root of two is 14.14213562373095101065700873732566833496093750000000


### Printing to 100 decimal places ........


While the algorithm itself will calculate the square root of two, it will not print out 100 decimal places so easily. The number before the decimal point cannot be rounded though so I decided to try using a large multiple of 2. In the same way that the square root of 4 is 2, square root of 400 is 20, (or square root of 9 is 3 and the square root of 900 is 30). 
- The square root of 4 times 100 (400) is the same as 100 times the sqaure root of 4.
- The square root of 900 is the same as 100 times the square root of 9.

If I get the square root of a multiple of two and then divide by that multiple I might be able to increase the number of decimal places by capturing the whole number part of a multiple of 100 and then dividing by 100.
(I am using the math module here just to see what the square root of a multiple of 2 would look like)

So first try multiplying by 100, then square root of the result and divide by 100.
Can use the `**` to raise to power of in Python. See [python docs](https://docs.python.org/3/tutorial/introduction.html#using-python-as-a-calculator)

In [80]:
print(2**2) # 2 to power of 2
print(10**10) # 10 to power of 10 is 10000000000
print (10**100)
# 10 to power of 
10**10


4
10000000000
10000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000


10000000000

10 to the power of 2 is 100, 10 to the power of 10 is 10000000000
10 to the power of 100 is a big number! I can get the length of this by using the length function on a string. First need to convert into a string using the `str` function and then use the `len` function on the result of that or `len(str()` in one go.

In [84]:
print(len(str(10**10)))
print(len(str(10**100)))
print(len(str(10**1000)))

11
101
1001


In [25]:
print(math.sqrt(2))
print(math.sqrt(2*(10**10)))
print(math.sqrt(2*(10**100)))
print(math.sqrt(2*(10**200))) 



1.4142135623730951
141421.35623730952
1.414213562373095e+50
1.414213562373095e+100


#### Scientific notation
The above is returned is scientific notation so somehow need to convert this.
its also not long enough!
(I cannot use math.sqrt on `2*(10**100)` as I get an overflow error, `int too large to convert`. This is incorrect as I can do it now. I must have tried on a larger number like 2 times 10 to power of a much larger number. I can do it on 2 times `10**400` and even larger.

The number 2 multiplied by 10 to the power of 200 is `200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000`


In [48]:
import math
print(2*(10**300))
print(math.sqrt(2*(10**300)))


2000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
1.4142135623730951e+150


So this is getting closer to what I want but I don't want to print scientific notation.
going with the multiple of 100 theory, the approximate square root of 2 is 1.414213567 so I will multiply this by 10 to power of 100.
Taking my function defined previously `my_sqrt` and trying it out a multiple of 2 times 10 to power of 100.
This still returns scientific notation. 

In [49]:
x =2*(10**100)
print(x)
print(len(str(x)))
y =1.414213567*(10**100)
print(y)

20000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
101
1.414213567e+100


In [54]:
my_sqrt(2*(10**200))

1.414213562373095e+100

So I can print out a long integer such as `x =2*(10**100)` or even 2 times 10 to power of 1000. I'm sure even more. 

In [56]:
x =2*(10**100)
print(x)
x =2*(10**1000)
print(x)
print(len(str(x)))

20000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
2000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000

So I still need to find a way of printing the scientific notation to float.
Look back again at [python docs on floating points numbers.](https://docs.python.org/3/tutorial/floatingpoint.html) which goes into detail as to why
>Unfortunately, most decimal fractions cannot be represented exactly as binary fractions. A consequence is that, in general, the decimal floating-point numbers you enter are only approximated by the binary floating-point numbers actually stored in the machine.

The docs do mention using string formatting to print a limited number of significant digits.
The example from the python docs shows how `.2f` is used to give 2 digits after the decimal point and `.12g` to give 12 significant digits.
I could specify `.100g`or`.100f`

### Trying the square root function on a large multiple of 2
I did try this earlier on the algorithm to find the square root of 2 but it printed zeros after the first 50 places so I next tried it on the algorithm applied to the multiple of 2 (2 multiplied by 10 to power of 100).
This produces the number `1.414213562373095027142412563281858698349164881791987548177900388860130684271654303022821004349852877e+100` which is in scientific notation. So now I have the digits but they are for the square root of a multple of two.
It might not even be accurate. Wikipedia gives the square root of 2 to 50 decimal places. (`1.41421356237309504880168872420969807856967187537694` which differs after the first 16 places! Oh dear.
The task is to print the square root of 2 to 100 decimal places. My answer here is the square root of 2 times 10 to power of 200. But it is a string and it is over 100 decimal places so I can work with the string.


In [99]:
x = 2*(10**200)
result =my_sqrt(x)
x100 =format(result,'.100g')
print(x100)
print(len(str(x100)))
print((1.414213562373095027142412563281858698349164881791987548177900388860130684271654303022821004349852877e+100)**2)


1.414213562373095027142412563281858698349164881791987548177900388860130684271654303022821004349852877e+100
106
2e+200


I square the result to see if it returns the original number that was passed to the `my_sqrt()` function and it returns`2e+200`.  `(1.414213562373095027142412563281858698349164881791987548177900388860130684271654303022821004349852877e+100)**2` 
I thought the last time I did this it returned 1.9999999999999996. See below where
`1.414213562373095027142412563281858698349164881791987548177900388860130684271654303022821004349852877**2` returned 1.9999999999999996.

In [89]:
# squaring the result 
(1.414213562373095027142412563281858698349164881791987548177900388860130684271654303022821004349852877e+100)**2

2e+200

In [94]:
## from before.
1.414213562373095027142412563281858698349164881791987548177900388860130684271654303022821004349852877**2

1.9999999999999996

### Extracting the digits from the string
I now need to take away the `e+200` from the end of the string.
See [strings tutorial](https://docs.python.org/3/tutorial/introduction.html#strings) from the Python docs. I can just take the first 102 characters including the 1 and the decimal point and not including the `e+200` 

In [95]:
# extracting the first 102 characters from the string `x100`
ans =x100[0:101]
print(ans)

1.414213562373095027142412563281858698349164881791987548177900388860130684271654303022821004349852877


In [96]:
# converting to float and squaring gives 1.9999999999999996
float(ans)*float(ans)

1.9999999999999996

In [101]:
# taking out the digits from the string and squaring:
1.414213562373095027142412563281858698349164881791987548177900388860130684271654303022821004349852877**2

1.9999999999999996

In [102]:
# squaring the actual result from the function:
(1.414213562373095027142412563281858698349164881791987548177900388860130684271654303022821004349852877e+100)**2

2e+200

In [262]:
math.sqrt(2)*math.sqrt(2)

2.0000000000000004

In [263]:
np.sqrt(2)*np.sqrt(2)

2.0000000000000004

So it looks like the number itself is correct but when converted to a string to get the decimals and then converted back to a float it gives a different number. Which is to be expected from something I read before!

# References
- https://www.latex-tutorial.com/tutorials/amsmath/
- https://www.math-linux.com/latex-26/faq/latex-faq/article/latex-natural-numbers

Read https://kodlogs.com/blog/715/convert-scientific-notation-to-decimal-python